pvlib/iotools/psm4.py

```diff
@@ -0,0 +1,388 @@
+        querying data with < 30-minute resolution uses a different API endpoint
+        with fewer available fields (see [4]_).
+
+    See Also
+    --------
+    pvlib.iotools.read_psm4, pvlib.iotools.parse_psm4
+
+    References
+    ----------
+
+    .. [1] `NREL National Solar Radiation Database (NSRDB)
+       <https://nsrdb.nrel.gov/>`_
+    .. [2] `NSRDB GOES Aggregated V4.0.0
+       <https://developer.nrel.gov/docs/solar/nsrdb/nsrdb-GOES-aggregated-v4-0-0-download/>`_
+    .. [3] `NSRDB GOES Tmy V4.0.0
+       <https://developer.nrel.gov/docs/solar/nsrdb/nsrdb-GOES-tmy-v4-0-0-download/>`_
+    .. [4] `NSRDB GOES Conus V4.0.0
+       <https://developer.nrel.gov/docs/solar/nsrdb/nsrdb-GOES-conus-v4-0-0-download/>`_
+    """
+    # The well know text (WKT) representation of geometry notation is strict.
+    # A POINT object is a string with longitude first, then the latitude, with
+    # four decimals each, and exactly one space between them.
+    longitude = ('%9.4f' % longitude).strip()
+    latitude = ('%8.4f' % latitude).strip()
+    # TODO: make format_WKT(object_type, *args) in tools.py
+
+    # convert to string to accomodate integer years being passed in
+    names = str(names)
+
+    # convert pvlib names in attributes to PSM4 convention
+    attributes = [REQUEST_VARIABLE_MAP.get(a, a) for a in attributes]
+
+    # required query-string parameters for request to PSM4 API
+    params = {
+        'api_key': api_key,
+        'full_name': full_name,
+        'email': email,
+        'affiliation': affiliation,
+        'reason': PVLIB_PYTHON,
+        'mailing_list': 'false',
+        'wkt': 'POINT(%s %s)' % (longitude, latitude),
+        'names': names,
+        'attributes':  ','.join(attributes),
+        'leap_day': str(leap_day).lower(),
+        'utc': 'false',
+        'interval': interval
+    }
+    # request CSV download from NREL PSM4
+    if url is None:
+        # determine the endpoint that suits the user inputs
+        if any(prefix in names for prefix in ('tmy', 'tgy', 'tdy')):
+            url = TMY_URL
+        elif interval in (5, 15):
+            url = GOESCONUS_URL
+        else:
+            url = GOES_URL
+
+    response = requests.get(url, params=params, timeout=timeout)
+    if not response.ok:
+        # if the API key is rejected, then the response status will be 403
+        # Forbidden, and then the error is in the content and there is no JSON
+        try:
+            errors = response.json()['errors']
+        except JSONDecodeError:
+            errors = response.content.decode('utf-8')
+        raise requests.HTTPError(errors, response=response)
+    # the CSV is in the response content as a UTF-8 bytestring
+    # to use pandas we need to create a file buffer from the response
+    fbuf = io.StringIO(response.content.decode('utf-8'))
+    return parse_psm4(fbuf, map_variables)
+
+
+def parse_psm4(fbuf, map_variables=True):
```

Comment:
I personally think we should get rid of the parse functions as it creates a LOT of duplicate documentation and significantly more functions than needed. It's unnecessary clutter and it takes very little to modify _read_ functions to also take filebuffers.

Alternatively, we could consider making the _parse_ functions private.

Reply:
I am generally +1 to ditching `parse_` and having `read_` work with buffers as well as filenames.  But let's do it as a follow-up PR.